### DISCUSSION TITLE
How to tilt the propeller using [uns.KinematicManeuver(angles, RPMs, Vvehicle, anglevehicle)] ?

### DISCUSSION THREAD

**Discussion by inse0918** [byuflowlab/FLOWUnsteady#63] [Q&A] [answered]
Hi, all.

I had tilt the isolated wing using [uns.KinematicManeuver(angles, RPMs, Vvehicle, anglevehicle)].

Like this,

    # ------------- 2) MANEUVER DEFINITION -----------------------------------------
    
    # Non-dimensional translational velocity of vehicle over time
    Vvehicle(t) = [-1, 0, 0]        # <---- Vehicle is traveling in the -x direction
    
    # Angle of the vehicle over time
    anglevehicle(t) = zeros(3)
    
    # Control inputs
    angle_frontwing(t) = [0, (900 * t*ttot), 0]  # Tilt angle of front wing
    
    angle = (angle_frontwing, )                 # Angle of each tilting system
    RPM = ()                                    # RPM of each rotor system (none)
    
    maneuver = uns.KinematicManeuver(angle, RPM, Vvehicle, anglevehicle)

Then the wing tilt(from 0 deg to 90 deg).

![image](https://user-images.githubusercontent.com/93040154/235589771-d2204de9-c1aa-4d97-9498-9ea80fbdb4c5.png)

So, I thought that it can be applied to tilt the propeller or rotor.

I fixed the code like this in 'propeller.jl',

    # ------------- 2) MANEUVER DEFINITION -----------------------------------------
    # Non-dimensional translational velocity of vehicle over time
    Vvehicle(t) = zeros(3)
    
    # Angle of the vehicle over time
    anglevehicle(t) = zeros(3)
    
    # RPM control input over time (RPM over `RPMref`)
    RPMcontrol(t) = 1.0
    angle_frontwing(t) = [0, (900 * t*ttot), 0]  # Tilt angle of front wing
    
    angles = (angle_frontwing, )                                 # Angle of each tilting system (none)
    RPMs = (RPMcontrol, )                       # RPM of each rotor system
    
    maneuver = uns.KinematicManeuver(angles, RPMs, Vvehicle, anglevehicle)

But, uns.Simulation do not apply this code.

    simulation = uns.Simulation(vehicle, maneuver, Vref, RPMref, ttot;
                                                        Vinit=Vinit, Winit=Winit);
    
    Encountered incompatible Vehicle and Maneuver!
    
    Stacktrace:
     [1] error(s::String)
       @ Base [./error.jl:35](https://vscode-remote+wsl-002bubuntu-002d22-002e04.vscode-resource.vscode-cdn.net/home/inse0918/.julia/packages/FLOWUnsteady/Jl5A6/examples/wing/error.jl:35)
     [2] _check(vehicle::FLOWUnsteady.UVLMVehicle{0, 1, Float64}, maneuver::FLOWUnsteady.KinematicManeuver{1, 1}; raise_error::Bool)
       @ FLOWUnsteady [~/.julia/packages/FLOWUnsteady/Jl5A6/src/FLOWUnsteady_simulation_types.jl:208](https://vscode-remote+wsl-002bubuntu-002d22-002e04.vscode-resource.vscode-cdn.net/home/inse0918/.julia/packages/FLOWUnsteady/Jl5A6/examples/wing/~/.julia/packages/FLOWUnsteady/Jl5A6/src/FLOWUnsteady_simulation_types.jl:208)
     [3] _check(vehicle::FLOWUnsteady.UVLMVehicle{0, 1, Float64}, maneuver::FLOWUnsteady.KinematicManeuver{1, 1})
       @ FLOWUnsteady [~/.julia/packages/FLOWUnsteady/Jl5A6/src/FLOWUnsteady_simulation_types.jl:202](https://vscode-remote+wsl-002bubuntu-002d22-002e04.vscode-resource.vscode-cdn.net/home/inse0918/.julia/packages/FLOWUnsteady/Jl5A6/examples/wing/~/.julia/packages/FLOWUnsteady/Jl5A6/src/FLOWUnsteady_simulation_types.jl:202)
     [4] FLOWUnsteady.Simulation{FLOWUnsteady.UVLMVehicle{0, 1, Float64}, FLOWUnsteady.KinematicManeuver{1, 1}, Float64}(vehicle::FLOWUnsteady.UVLMVehicle{0, 1, Float64}, maneuver::FLOWUnsteady.KinematicManeuver{1, 1}, Vref::Float64, RPMref::Int64, ttot::Float64; Vinit::Vector{Float64}, Winit::Vector{Float64}, t::Float64, nt::Int64)
       @ FLOWUnsteady [~/.julia/packages/FLOWUnsteady/Jl5A6/src/FLOWUnsteady_simulation_types.jl:58](https://vscode-remote+wsl-002bubuntu-002d22-002e04.vscode-resource.vscode-cdn.net/home/inse0918/.julia/packages/FLOWUnsteady/Jl5A6/examples/wing/~/.julia/packages/FLOWUnsteady/Jl5A6/src/FLOWUnsteady_simulation_types.jl:58)
     [5] FLOWUnsteady.Simulation(::FLOWUnsteady.UVLMVehicle{0, 1, Float64}, ::FLOWUnsteady.KinematicManeuver{1, 1}, ::Float64, ::Int64, ::Vararg{Any}; optargs::Base.Pairs{Symbol, Vector{Float64}, Tuple{Symbol, Symbol}, NamedTuple{(:Vinit, :Winit), Tuple{Vector{Float64}, Vector{Float64}}}})
       @ FLOWUnsteady [~/.julia/packages/FLOWUnsteady/Jl5A6/src/FLOWUnsteady_simulation_types.jl:72](https://vscode-remote+wsl-002bubuntu-002d22-002e04.vscode-resource.vscode-cdn.net/home/inse0918/.julia/packages/FLOWUnsteady/Jl5A6/examples/wing/~/.julia/packages/FLOWUnsteady/Jl5A6/src/FLOWUnsteady_simulation_types.jl:72)
     [6] top-level scope
       @ [~/.julia/packages/FLOWUnsteady/Jl5A6/examples/wing/test.ipynb:11](https://vscode-remote+wsl-002bubuntu-002d22-002e04.vscode-resource.vscode-cdn.net/home/inse0918/.julia/packages/FLOWUnsteady/Jl5A6/examples/wing/~/.julia/packages/FLOWUnsteady/Jl5A6/examples/wing/test.ipynb:11)

Is there any way to tilt the propeller or rotor ?

Thank you.
Best regards,
Inseo

**Comment by inse0918** [comment, marked as answer]
Hi, all.

When you want to tilt the propeller angle over the increasing time, you should handle the **anglevehicle(t)**

The example code about **heavingwing.jl** the tilt angle variable is

    # Control inputs
    angle_frontwing(t) = [0, amplitude*sin(2*pi*frequency * t*ttot), 0]  # Tilt angle of front wing

But, this control input does not apply in propeller code(IDK why it is,,,)

But you can handle the tilting angle of the propeller using variable **anglevehicle(t)**

Like this,

    # Angle of the vehicle over time
    anglevehicle(t) = [0, (900 * t*ttot), 0]

Then the propeller tilting the angle from the 0 deg to 90 deg

![image](https://user-images.githubusercontent.com/93040154/235601509-5fc31554-8203-48f8-896b-327e8b16f8f5.png)

Thank you.
Best regards,
Inseo

**Comment by EdoAlvarezR** [comment]
I'm glad you were able to get your simulation running @inse0918, and thanks for sharing your journey!

The `Encountered incompatible Vehicle and Maneuver!` error comes from defining a maneuver that doesn't match either the same number of rotor systems or tilting systems as defined in the vehicle. It's difficult to tell without looking at your Vehicle definition, but my best guess is that you didn't add the rotor as a tilting system. You found a workaround by tilting the entire vehicle instead.

**Comment by EdoAlvarezR** [comment]
Also, you might find the [Visualization guide](https://flow.byu.edu/FLOWUnsteady/visualization) in the docs insightful as you go about processing your simulations =] 

https://flow.byu.edu/FLOWUnsteady/visualization

**Reply by inse0918** [reply]
Thank you for pointing tip of the post-processing !
